### PR TITLE
Add typescript definitions

### LINKED ISSUE #22: Add Typescript typedefs
There's been some interest in Typescript type definitions for formula-one. It should be relatively easy to make them.

### PR DESCRIPTION
**Pull request by zgotsch**
Fixes #22

### CODE REVIEW

**Comment by uforic** [comment]
image

Thanks chief!

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  export const Touched: FeedbackStrategy;
+  export const Changed: FeedbackStrategy;
+  export const ClientValidationSucceeded: FeedbackStrategy;
+  export const Pristine: FeedbackStrategy;
+  export const Submitted: FeedbackStrategy;
+}
+
+export type ExternalErrors = null | {[path: string]: ReadonlyArray<string>};
+
+export type FieldLink<T> = {__type__: "FieldLink"};
```
Comment (dmnd) [inline index.d.ts]: "opaque", huh

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  validation?: Validation<T>;
+  children: (
+    value: T,
+    errors: ReadonlyArray<string>,
+    onChange: (value: T) => void,
+    onBlur: () => void,
+    additionalInfo: AdditionalRenderInfo<T>
+  ) => React.ReactNode;
+}
+declare class Field<T> extends React.Component<T> {}
```
[the remark this answers is not in the document]
Comment (zgotsch) [reply index.d.ts]: Oops, let me look into why I didn't get any type errors about this.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+export type Span<T> = [number, ReadonlyArray<T>];
+export type AddField<T> = (index: number, value: T) => void;
+export type RemoveField = (index: number) => void;
+export type MoveField = (oldIndex: number, newIndex: number) => void;
+export type AddFields<T> = (spans: ReadonlyArray<Span<T>>) => void;
+export type FilterFields<T> = (
+  predicate: (value: T, index: number, arr: ReadonlyArray<T>) => boolean
+) => void;
+export type ModifyFields<T> = (
```
[the remark this answers is not in the document]
Comment (zgotsch) [reply index.d.ts]: They should be exported as Flow types too. See #30 for context.

File: index.d.ts (hunk 1 above)
Comment (zgotsch) [reply index.d.ts]: yep!

**Comment by zgotsch** [comment]
The test plan was I built a simple CRA which used formula-one and verified I was getting reasonable type errors.

There is a somewhat interesting question of whether we should host the type definitions ourself as here or put them on DefinitelyTyped so that they would be included in a separate package. Pros and cons discussed: https://stackoverflow.com/questions/56834190/why-publish-the-typescript-declaration-file-on-definitelytyped-for-a-javascript

File: index.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (zgotsch) [reply index.d.ts]: Didn't get any errors because Typescript secretly made the props `any`. Thanks, Typescript 😢

**Pull request merged**